evdns.c

```diff
@@ -742,9 +742,32 @@ nameserver_failed(struct nameserver *const ns, const char *msg) {
 	ns->state = 0;
 	ns->failed_times = 1;
 
-	disconnect_and_free_connection(ns->connection);
-	ns->connection = NULL;
+	if (ns->connection) {
+		disconnect_and_free_connection(ns->connection);
+		ns->connection = NULL;
+	} else if (err == ENOTCONN) {
+		// XXX: https://github.com/libevent/libevent/issues/265
+		const struct sockaddr *address = (const struct sockaddr *)&ns->address;
+		evutil_closesocket(ns->socket);
+		ns->socket = evutil_socket_(address->sa_family,
+			SOCK_DGRAM | EVUTIL_SOCK_NONBLOCK | EVUTIL_SOCK_CLOEXEC, 0);
+
+		if (base->global_outgoing_addrlen &&
+			!evutil_sockaddr_is_loopback_(address)) {
+			if (bind(ns->socket,
+					(struct sockaddr *)&base->global_outgoing_address,
+					base->global_outgoing_addrlen) < 0) {
+				log(EVDNS_LOG_WARN, "Couldn't bind to outgoing address");
+			}
+		}
 
+		event_del(&ns->event);
+		event_assign(&ns->event, ns->base->event_base, ns->socket,
+			EV_READ | (ns->write_waiting ? EV_WRITE : 0) | EV_PERSIST,
+			nameserver_ready_callback, ns);
+		if (!base->disable_when_inactive && event_add(&ns->event, NULL) < 0) {
```

Comment:
Please cleanup the code, empty if branch is not ok, let's add log message at least

Reply:
fixed